### PR TITLE
feat: [#168748555] Force CVV insertion during MAESTRO payments

### PR DESCRIPTION
**Pull request by matgentili**
In this pr:
- restored Maestro card
- now during payment with MAESTRO card is necessary to insert the verification cvc

With this pr https://github.com/teamdigitale/io-app/pull/1431 the screenHeader will be fixed

Screenshot_20191121-143916

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#168748555](https://www.pivotaltracker.com/story/show/168748555): Come CIT se, durante il pagamento, scelgo di pagare con una carta MAESTRO devo necessariamente inserire il CVC prima di procedere

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please include a Pivotal story at the beginning of the PR title (see below).
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Remember to fix the PR title by removing WIP wording when ready
  </td>
    </tr>
  </tbody>
</table>



  Example of PR titles that include pivotal stories:

  * single story: `[#123456] my PR title`
  * multiple stories: `[#123456,#123457,#123458] my PR title`

    
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please include a Pivotal story at the beginning of the PR title (see below).
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Remember to fix the PR title by removing WIP wording when ready
  </td>
    </tr>
  </tbody>
</table>



  Example of PR titles that include pivotal stories:

  * single story: `[#123456] my PR title`
  * multiple stories: `[#123456,#123457,#123458] my PR title`

    
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#168748555](https://www.pivotaltracker.com/story/show/168748555): Come CIT se, durante il pagamento, scelgo di pagare con una carta MAESTRO devo necessariamente inserire il CVC prima di procedere

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: ts/utils/payment.ts [hunk 1]
```diff
@@ -122,3 +122,10 @@ export const cleanTransactionDescription = (description: string): string => {
     ? descriptionParts[descriptionParts.length - 1].trim()
     : "";
 };
+
+/**
+ * This function control if card brand is MAESTRO.
+ */
+export const isBrandMaestro = (card: CreditCard): boolean => {
+  return card.brand === "MAESTRO";
```
Comment (Undermaken) [inline ts/utils/payment.ts]: if brand is "maestro" it returns false.
An insensitive-case comparison could be better

**Comment by Undermaken** [comment]
@valentinacala  update!

File: ts/components/ChooserListContainer.tsx [hunk 2]
```diff
@@ -294,7 +294,7 @@ class ChooserListContainer<T> extends React.PureComponent<Props<T>, State> {
           </View>
         </Content>
         <KeyboardAvoidingView
-          behavior={Platform.OS === "android" ? "height" : "padding"}
+          behavior={"padding"}
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/ChooserListContainer.tsx]: reverted within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+import TopScreenComponent from "./screens/TopScreenComponent";
+import FooterWithButtons from "./ui/FooterWithButtons";
+import IconFont from "./ui/IconFont";
+import MaskedInput from "./ui/MaskedInput";
+import PaymentBannerComponent from "./wallet/PaymentBannerComponent";
+
+type Props = Readonly<{
+  onCancel: () => void;
+  onContinue: (cvv?: string) => void;
+  verifica: any;
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+  public componentDidMount() {
+    BackHandler.addEventListener("hardwareBackPress", this.handleBackPress);
+  }
+
+  public componentWillUnmount() {
+    BackHandler.removeEventListener("hardwareBackPress", this.handleBackPress);
+  }
+
+  private updateSecurityCodeState(value: string) {
+    this.setState({
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+      securityCode:
+        value && value !== EMPTY_CARD_SECURITY_CODE ? some(value) : none
+    });
+  }
+  private isValidSecurityCode() {
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            keyboardShouldPersistTaps={"handled"}
+          >
+            <ScreenContentHeader
+              title={I18n.t("wallet.confirmPayment.insertCode")}
+              icon={require("./../../img/wallet/cvc-icon.png")}
+            />
+            <Content scrollEnabled={false}>
+              <LabelledItem
+                type={"masked"}
+                isValid={this.isValidSecurityCode()}
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+          )}
+        </KeyboardAvoidingView>
+      </TopScreenComponent>
+    );
+  }
+
+  /**
+   * Footer
+   */
+  private renderFooterButtons(isValidCode: boolean) {
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+      light: true,
+      bordered: true,
+      onPress: this.props.onCancel,
+      cancel: true,
+      title: I18n.t("global.buttons.cancel")
+    };
+    const primaryButtonProps = {
+      block: true,
+      primary: true,
+      disabled: isValidCode,
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        </KeyboardAvoidingView>
+      </TopScreenComponent>
+    );
+  }
+
+  /**
+   * Footer
+   */
+  private renderFooterButtons(isValidCode: boolean) {
+    const handleOnContinue = () => {
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/utils/payment.ts (hunk 1 above)
Comment (valentinacala) [reply ts/utils/payment.ts]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

**Comment by valentinacala** [comment]
@matteodesanti wip until a maestro card can be used on production environment

File: ts/components/PaymentSecureCodeOverlay.tsx [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+          </ScrollView>
+        </Content>
+        <KeyboardAvoidingView
+          behavior={"padding"}
+          keyboardVerticalOffset={Platform.select({
+            ios: 0,
+            android: variables.contentPadding
+          })}
+        >
+          {this.renderFooterButtons(
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: updated within https://github.com/pagopa/io-app/pull/1351/commits/2771838cffb307cfa630a373dd7734dcbe77aa71

File: ts/components/PaymentSecureCodeOverlay.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/components/PaymentSecureCodeOverlay.tsx]: cool function! Updated within https://github.com/pagopa/io-app/pull/1351/commits/5f02b771e4f109f62fbd20807bbc1dd5198951a7

File: android/app/src/main/AndroidManifest.xml [hunk 11]
```diff
@@ -40,7 +40,7 @@
     </service>
     <!-- END Required by react-native-push-notification -->
 
-    <activity android:name=".MainActivity" android:launchMode="singleTask" android:label="@string/app_name" android:configChanges="keyboard|keyboardHidden|orientation|screenSize" android:windowSoftInputMode="adjustResize">
+    <activity android:name=".MainActivity" android:launchMode="singleTask" android:label="@string/app_name" android:configChanges="keyboard|keyboardHidden|orientation|screenSize" android:windowSoftInputMode="adjustPan">
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply android/app/src/main/AndroidManifest.xml]: I don't remember neither. 
 https://github.com/pagopa/io-app/pull/1351/commits/5f02b771e4f109f62fbd20807bbc1dd5198951a7 I checked the screen with the input and the other similar screens (such as AddCardScreen), reverted this change and update the screen to have a more coherent design if compared with the invisio.
Here a comparison:
image

File: ts/screens/wallet/AddCardScreen.tsx [hunk 12]
```diff
@@ -378,9 +379,11 @@ class AddCardScreen extends React.Component<Props, State> {
   }
 
   private updateSecurityCodeState(value: string) {
+    const securityCode = fromNullable(value).filter(
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/screens/wallet/AddCardScreen.tsx]: updated ( https://github.com/pagopa/io-app/pull/1351/commits/5f02b771e4f109f62fbd20807bbc1dd5198951a7)

File: ts/utils/payment.ts [hunk 13]
```diff
@@ -130,3 +130,14 @@ export const cleanTransactionDescription = (description: string): string => {
     description.length > maxLength ? "..." : ""
   }`;
 };
+
+/**
+ * This function control if card brand is MAESTRO.
+ */
+export const isBrandMaestro = (card: CreditCard): boolean => {
```
[the remark this answers is not in the document]
Comment (valentinacala) [reply ts/utils/payment.ts]: updated( https://github.com/pagopa/io-app/pull/1351/commits/5f02b771e4f109f62fbd20807bbc1dd5198951a7)

File: android/app/src/main/AndroidManifest.xml (hunk 11 above)
[the remark this answers is not in the document]
Comment (Undermaken) [reply android/app/src/main/AndroidManifest.xml]: @valentinacala 
So we should keep this change?

[the remark this answers is not in the document]
Comment (valentinacala) [reply android/app/src/main/AndroidManifest.xml]: @Undermaken nope, I reverted it since with the new code works as expected even without it.

**Pull request closed without merging**